app/src/main/java/com/hedvig/app/feature/embark/passages/multiaction/add/AddComponentBottomSheet.kt

```diff
@@ -0,0 +1,155 @@
+package com.hedvig.app.feature.embark.passages.multiaction.add
+
+import android.os.Bundle
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import android.widget.AdapterView
+import android.widget.ArrayAdapter
+import androidx.appcompat.widget.SwitchCompat
+import androidx.core.os.bundleOf
+import androidx.core.widget.doOnTextChanged
+import androidx.fragment.app.setFragmentResult
+import com.google.android.material.bottomsheet.BottomSheetDialogFragment
+import com.google.android.material.textfield.TextInputLayout
+import com.hedvig.app.R
+import com.hedvig.app.databinding.DialogAddBuildingBinding
+import com.hedvig.app.databinding.LayoutComponentDropdownBinding
+import com.hedvig.app.databinding.LayoutComponentNumberBinding
+import com.hedvig.app.databinding.LayoutComponentSwitchBinding
+import com.hedvig.app.feature.embark.passages.multiaction.MultiActionComponent
+import com.hedvig.app.feature.embark.passages.multiaction.MultiActionItem
+import com.hedvig.app.feature.embark.passages.multiaction.MultiActionParams
+import com.hedvig.app.util.extensions.hideKeyboard
+import com.hedvig.app.util.extensions.onImeAction
+import com.hedvig.app.util.extensions.viewBinding
+import org.koin.androidx.viewmodel.ext.android.viewModel
+import org.koin.core.parameter.parametersOf
+
+class AddComponentBottomSheet : BottomSheetDialogFragment() {
+
+    private val componentState: MultiActionItem.Component? by lazy {
+        requireArguments().getParcelable(COMPONENT_STATE)
+    }
+
+    private val multiActionParams: MultiActionParams by lazy {
+        requireArguments().getParcelable<MultiActionParams>(MULTI_ACTION_PARAMS)
+            ?: throw Error("Programmer error: No multi action params provided to ${this.javaClass.name}")
+    }
+
+    private val viewModel: AddComponentViewModel by viewModel { parametersOf(componentState, multiActionParams) }
+    private val binding by viewBinding(DialogAddBuildingBinding::bind)
+
+    override fun onCreateView(
+        inflater: LayoutInflater,
+        container: ViewGroup?,
+        savedInstanceState: Bundle?
+    ): View? = inflater.inflate(R.layout.dialog_add_building, container, false)
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+        multiActionParams.components.map { component ->
+            val componentView = when (component) {
+                is MultiActionComponent.Dropdown -> createDropDownComponent(component)
+                is MultiActionComponent.Number -> createNumberComponent(component)
+                is MultiActionComponent.Switch -> createSwitchComponent(component)
+            }
+
+            binding.componentContainer.addView(componentView)
+        }
+
+        viewModel.viewState.observe(this) {
```

Comment:
We need to use `viewLifecycleOwner` here, or we might get crashes if the `LiveData` emits while this fragment is within the `Fragment` lifecycle but outside the `Fragment` view-lifecycle. A real landmine 😞 

```suggestion
        viewModel.viewState.observe(viewLifecycleOwner) {
```